cql3/abstract_marker.cc

```diff
@@ -69,18 +69,19 @@ abstract_marker::raw::raw(int32_t bind_index)
 
 ::shared_ptr<term> abstract_marker::raw::prepare(database& db, const sstring& keyspace, lw_shared_ptr<column_specification> receiver) const
 {
-    if (receiver->type->is_collection()) {
-        if (receiver->type->get_kind() == abstract_type::kind::list) {
+    const auto tp = receiver->type->underlying_type();
```

Comment:
Wow. So potentially we have such type->get_kind() calls returning the wrong thing for reversed collections all over the code. Would be good if you or someone else could audit for that...
I'm curious - why can't type->get_kind() return the "kind" of the underlying type, to solve this problem for all callers?

Reply:
Also, a question: I understand what underlying_type does for reversed-order types. Is it possible that for other things, it has a different meaning, so it wouldn't be correct to use underlying_type?

Reply:
> Is it possible that for other things, it has a different meaning, so it wouldn't be correct to use underlying_type?

That's definitely possible, which is why I had to add the extra is_reversed() check in lists.cc.  I'll add it everywhere, just to be sure.

Reply:
I added two utility methods to abstract_type: self_or_reversed() and as().  But they must be called explicitly.